Add Support for Three Vector Scale Specification

liangfok opened the pull request:
Closes https://github.com/RobotLocomotion/drake/issues/2242

Updated SDF and URDF parsers to supports mesh scale factors specified as 3 vectors. Maintains support for mesh scale factors specified as a single scalar value. Scale specification using 3 vectors are part of the URDF / SDF standards.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2260)
<!-- Reviewable:end -->

liangfok commented:
The Win32 / Win64 CI build failures in [CI Build 1221](https://drake-jenkins.csail.mit.edu/job/experimental/1221/) appear to be due to https://github.com/RobotLocomotion/drake/issues/2257.  I will thus elevate this PR to a feature reviewer. @patmarion, since you originally implemented the feature in this PR, can you perform the feature review? I will go ahead assign this PR to you. Feel free to bounce it back to me if you're unable to perform the review.

liangfok commented:
Win32 / Win64 bug identified: https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=15957

I'm working on a fix for it now.

patmarion commented on drake/systems/plants/xmlUtil.cpp:
```diff
@@ -207,14 +241,14 @@ string resolveFilename(const string& filename,
-    bool dirIsRelative = !(normalized_root_dir.size() >= 1
-                           && normalized_root_dir[0] == '/');
+    bool dirIsRelative =
+        !(normalized_root_dir.size() >= 1 && normalized_root_dir[0] == '/');
```
I'm not a big fan of including unrelated style changes in pull requests.  It hides information from git blame, and makes it more difficult to review diffs.

liangfok replied:
I believe TRI requires running clang-format on all files touched. @david-german-tri is this correct?

patmarion replied:
yeah sorry I introduced that code and I did not run clang-format.  It did pass the CI style test, though, so I didn't realize I needed to run the format tool.

liangfok replied:
@david-german-tri  and @jwnimmer-tri: I think we need to investigate the 5 whys here. Why did @patmarion's change pass CI's `clang-format` check but not my local execution of it?

jwnimmer-tri replied:
CI only runs `cpplint`, not `clang-format`.  (I don't think it should run `clang-format`.)

david-german-tri replied:
Why?  CI doesn't have a `clang-format` check, only a `cpplint` check.
Why? `clang-format` isn't actually a requirement yet, just a recommended useful tool.  
Why? We don't trust it enough.  It's right 95% of the time but does some bad things in corner cases.

:-)

liangfok replied:
Ah. That explains it. Created https://github.com/RobotLocomotion/drake/issues/2265 to track this new issue.

david-german-tri replied:
Put another way: if your code is clang-format clean (and doesn't look like it fell off a corner case and into a volcano), it will also be style-guide compliant, saving us all some nitpicking.  

That said, clang-format will also change some style-guide compliant formatting into other style-guide compliant formatting, as seen here.  I think clang-format-using authors should feel free to make such changes.  It would be a big hassle to run clang-format and then back out its changes on lines you didn't otherwise affect.

liangfok replied:
In this case, should I surround @patmarion's originally-styled code with:

```
// clang-format off
...
// clang-format on
```
Comments?

jwnimmer-tri replied:
The complaint wasn't that the new format was poor, it was to keep reformat PRs and function PRs separate, for better git history.

david-german-tri replied:
I don't think so. As far as I can tell, no one has suggested @patmarion's formatting is superior: your fix is equally compliant, and has the advantage of being `clang-format` stable.

liangfok replied:
For this PR, should I undo the style changes and issue a separate PR with the `clang-format` compatible style changes?

patmarion replied:
let's keep the style change in this PR.

If I could go back I would have run clang-format on my code.  Since I didn't, it's my fault that git blame will be overwritten with a style formatting.  Not a big loss.

jwnimmer-tri replied:
To me, it seems a small enough change to not matter either way, and just treat it as a learning experience.

patmarion commented:
Looks fine to me.  I think the new function name parseThreeVectorAttribute is a little confusing.  It's the same as the existing parseVectorAttribute function, except it has the special behavior of detecting a scalar value and replicating it into a 3 vector, but the doc you added explains the behavior, so that's fine.  Thanks!

I don't think the parsing methods in xmlUtil are as robust as they could be, they could use more error checking, but that discussion doesn't need to take place here.

liangfok commented:
@david-german-tri I'm going to assign this to you for platform review / merge.

david-german-tri commented:
Face-to-face, @liangfok pointed out the `RigidBodyTree*` changes have test coverage from `drake/systems/plants/test/compareRigidBodySystems`

---

Reviewed 6 of 7 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdarKSEhcwtDll4IK:-KGxdarKSEhcwtDll4IL:949464503) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
Reference arguments must be const.  Please pass a pointer instead:

https://google.github.io/styleguide/cppguide.html#Reference_Arguments

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdq-IGXC3Iso7hWiq:-KGxdq-IGXC3Iso7hWir:1195399003) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
Please make new functions comply with the style guide on function names:

https://google.github.io/styleguide/cppguide.html#Function_Names

---

*[drake/systems/plants/xmlUtil.h, line 54 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxeQA1ynU-MX-xw2Nz:-KGxeQA1ynU-MX-xw2O-:1832290804) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.h#L54)):*
Please add some unit test coverage for these, so that the behavior is protected even if the mesh-scale-specific tests go away at some point in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KGxdNJZu6ydJK1Wm92V:-977329124)*
<!-- Sent from Reviewable.io -->

patmarion commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdarKSEhcwtDll4IK:-KGxfIlKQ6ssg0u9exnm:536036718) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
The new functions comply with the argument style of the existing functions in this file.  IMHO, that is more important, but I won't pick a fight :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdarKSEhcwtDll4IK:-KGxgIvgROoBVUuk0ju_:532717643) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
Sure.  Fixing the whole file in a follow-up PR would be a reasonable alternative.  See e.g. #1921 for precedent.  Ditto below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

liangfok commented:
I've addressed all reviewer comments. PTAL. Thanks!

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Face-to-face, @liangfok pointed out the `RigidBodyTree*` changes have test coverage from `drake/systems/plants/test/compareRigidBodySystems`</details>


---

Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdarKSEhcwtDll4IK:-KHmnSCar7WHB1TfQkpv:78234523) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Sure.  Fixing the whole file in a follow-up PR would be a reasonable alternative.  See e.g. #1921 for precedent.  Ditto below.</details>
Done. I modified the output parameters to be pointers to conform to the style guide.

---

*[drake/systems/plants/xmlUtil.cpp, line 15 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxdq-IGXC3Iso7hWiq:-KHmn_J7Rm-I39r6cZ0F:1088799204) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L15)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Please make new functions comply with the style guide on function names:
> 
> https://google.github.io/styleguide/cppguide.html#Function_Names</details>
Done. I used CamelCase with the first letter capitalized.

---

*[drake/systems/plants/xmlUtil.cpp, line 251 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGwPmNEPCLfffuD9Qap-r2-251:-KHmau9aOVGc2I3tpmwb:1299884446) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.cpp#L251)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> To me, it seems a small enough change to not matter either way, and just treat it as a learning experience.</details>
Can somebody with the necessary bits mark this thread as resolved? I will in the future try to separate style changes from functional changes...promise! :smile: 

---

*[drake/systems/plants/xmlUtil.h, line 54 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KGxeQA1ynU-MX-xw2Nz:-KHmnCR1wWs1PHiEbYcs:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/9a3ca5973f70ae56f080f386a7b0187c0562c523/drake/systems/plants/xmlUtil.h#L54)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Please add some unit test coverage for these, so that the behavior is protected even if the mesh-scale-specific tests go away at some point in the future.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KHrYEnJIUO7fyzo-Zdf:-1149404080)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed [CI Build 1444](https://drake-jenkins.csail.mit.edu/job/experimental/1444/) and is thus ready for review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I've addressed all reviewer comments. PTAL. Thanks!</details>


---

Review status: 2 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KHtOn37xkd0bq8poJJV:436565877)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@jwnimmer-tri for final platform review / merge.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed [CI Build 1444](https://drake-jenkins.csail.mit.edu/job/experimental/1444/) and is thus ready for review.
</details>


---

Review status: 2 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KI9npzka7DK3vD-mTAg:689426582)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 7 files at r1, 1 of 3 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 18 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAAy2g-YZf60YBrtVo:-KIAAy2g-YZf60YBrtVp:-1936995609) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L18)):*
> ```cpp
>   if (val == nullptr)
>     throw std::runtime_error(
>         "ERROR: ParseThreeVectorValue: parameter val is nullptr!");
> ```

Multi-line body requires braces.

---

*[drake/systems/plants/xmlUtil.cpp, line 21 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIABjy_4gWCayUx2VDo:-KIABjy_4gWCayUx2VDp:1577061501) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L21)):*
> ```cpp
>   if (strval) {
>     std::stringstream ss(strval);
>     // supports a single scalar value or 3-vector value.
> ```

Capitalize.

---

*[drake/systems/plants/xmlUtil.cpp, line 24 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAC66RlRVFnd_7P2iY:-KIAC66RlRVFnd_7P2iZ:-788155924) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L24)):*
> ```cpp
>     ss >> (*val)[0];
>     if (ss.good()) {
>       ss >> (*val)[1] >> (*val)[2];
> ```

If strval has two items, this appears to return `true`, but should be `false`.  Missing unit test for this scenario.

---

*[drake/systems/plants/xmlUtil.cpp, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAEjyIxmUF6oxfUYNM:-KIAEjyJU2Gcoxox-HmD:-189974055) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L29)):*
> ```cpp
>       (*val)[2] = (*val)[0];
>     }
>     return true;
> ```

Should a strval with more than three items be a parse error?

---

*[drake/systems/plants/xmlUtil.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAEzNhjIxR2KyFrGGz:-KIAEzNhjIxR2KyFrGH-:337838016) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L31)):*
> ```cpp
>     return true;
>   } else {
>     return false;
> ```

Missing unit test case for this branch.

---

*[drake/systems/plants/xmlUtil.cpp, line 39 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAF2xRaUbiH3QFrDvU:-KIAF2xRaUbiH3QFrDvV:337838016) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L39)):*
> ```cpp
>     return ParseThreeVectorValue(node->FirstChild()->Value(), val);
>   else
>     return false;
> ```

Missing unit test case for this branch.

---

*[drake/systems/plants/xmlUtil.cpp, line 42 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAF9_tlK4Gg5E9NE4C:-KIAF9_tlK4Gg5E9NE4D:428106685) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L42)):*
> ```cpp
> }
> 
> bool ParseThreeVectorValue(tinyxml2::XMLElement* node, const char* element_name,
> ```

Isn't it weird that other functions treat a null node or strval as a parse error, but this one segfaults?

---

*[drake/systems/plants/xmlUtil.cpp, line 50 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAFF7T_MVvTFctLjl1:-KIAFF7T_MVvTFctLjl2:-1576849725) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L50)):*
> ```cpp
>                                const char* attribute_name,
>                                Eigen::Vector3d* val) {
>   return ParseThreeVectorValue(node->Attribute(attribute_name), val);
> ```

Isn't it weird that other functions treat a null or strval node as a parse error, but this one segfaults?

---

*[drake/systems/plants/xmlUtil.h, line 5 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA7UggE7dYftVPu8cF:-KIA7Ugh071s6Qfi9DRU:-295708020) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L5)):*
> ```objc
> #include <Eigen/Dense>
> #include <map>
> #include <string>
> ```

These two are in the wrong place now.

---

*[drake/systems/plants/xmlUtil.h, line 10 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA7azIAPvzZtPl3IXW:-KIA7azIAPvzZtPl3IXX:754049871) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L10)):*
> ```objc
> #include "drake/systems/plants/pose_map.h"
> #include "drake/thirdParty/tinyxml2/tinyxml2.h"
> #include "spruce.hh"
> ```

This one is in the wrong place now.

---

*[drake/systems/plants/xmlUtil.h, line 44 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA8PLaI3OrEz5BnXWv:-KIA8PLaI3OrEz5BnXWw:1459644117) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L44)):*
> ```objc
> 
> /**
>  * Parses a three vector value from parameter \p strval. If \p strval points to
> ```

API doc fails to explain or cite the syntax(es) of the strval that will be successfully parsed.

---

*[drake/systems/plants/xmlUtil.h, line 52 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADOJA_nQB-1gwBe8a:-KIADOJBROZMP9mgO4oe:-1672490036) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L52)):*
> ```objc
>  * @param[out] val The three vector into which the results should be stored.
>  * @return Whether the three vector was successfully parsed from \p strval.
>  * @throws std::runtime_error If parameter \p val is nullptr.
> ```

Missing unit test for this API claim.  (Perhaps just remove the API claim?)

---

*[drake/systems/plants/xmlUtil.h, line 67 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADScpvUOFvkVOAEkx:-KIADScsCJf5YfmpfCPR:-1672490036) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L67)):*
> ```objc
>  * @return Whether the three vector was successfully parsed from the XML element
>  * node.
>  * @throws std::runtime_error If parameter \p val is nullptr.
> ```

Missing unit test for this API claim.  (Perhaps just remove the API claim?)

---

*[drake/systems/plants/xmlUtil.h, line 70 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA9yuxlpX1q__o5drY:-KIA9yuyfyRj2HS78U0n:733000351) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L70)):*
> ```objc
>  */
> DRAKEXMLUTIL_EXPORT
> bool ParseThreeVectorValue(tinyxml2::XMLElement* node, Eigen::Vector3d* val);
> ```

`const tinyxml2::XMLElement*` type for input

---

*[drake/systems/plants/xmlUtil.h, line 86 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADU3ZHcGm1ZJmpdkP:-KIADU3ZHcGm1ZJmpdkQ:-1672490036) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L86)):*
> ```objc
>  * @return Whether the three vector was successfully parsed from the XML element
>  * node.
>  * @throws std::runtime_error If parameter \p val is nullptr.
> ```

Missing unit test for this API claim.  (Perhaps just remove the API claim?)

---

*[drake/systems/plants/xmlUtil.h, line 89 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAA8qMV03baXGC3Pty:-KIAA8qMV03baXGC3Ptz:919804796) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L89)):*
> ```objc
>  */
> DRAKEXMLUTIL_EXPORT
> bool ParseThreeVectorValue(tinyxml2::XMLElement* node, const char* element_name,
> ```

`const ...XMLElement*` type for input

---

*[drake/systems/plants/xmlUtil.h, line 107 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADUmJneoG1P9JjXIY:-KIADUmJneoG1P9JjXIZ:-1672490036) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L107)):*
> ```objc
>  * @return Whether the three vector was successfully parsed from the XML element
>  * node.
>  * @throws std::runtime_error If parameter \p val is nullptr.
> ```

Missing unit test for this API claim.  (Perhaps just remove the API claim?)

---

*[drake/systems/plants/xmlUtil.h, line 110 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAA9ZGBX-E4nwID4D1:-KIAA9ZGBX-E4nwID4D2:919804796) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L110)):*
> ```objc
>  */
> DRAKEXMLUTIL_EXPORT
> bool ParseThreeVectorAttribute(tinyxml2::XMLElement* node,
> ```

`const ...XMLElement*` type for input

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 30 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIACiKodr2iAWr1IkUH:-KIACiKps4_0gA79dDAC:-59262876) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/test/testXmlUtil.cpp#L30)):*
> ```cpp
> // the case where the string describes a three vector and the case where
> // the string describes a single scalar value.
> GTEST_TEST(test_parse_three_vector, from_string) {
> ```

Is there a citation for why this first argument should be different than the filename, and different within a file?

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 34 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIACuGD9Vpj1B9wloVE:-KIACuGD9Vpj1B9wloVF:-2118585860) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/test/testXmlUtil.cpp#L34)):*
> ```cpp
>   const char* one_value_string = "4.4";
> 
>   Vector3d parsed_vector3d;
> ```

BTW If Eigen weren't so fabulously verbose for how to zero things, it would be a defect for lack of initializing a variable at its point of declaration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@jwnimmer-tri for final platform review / merge.
</details>


---

Review status: 5 of 8 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[drake/systems/plants/xmlUtil.cpp, line 18 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAAy2g-YZf60YBrtVo:-KIDjZCx6FeKverxpMur:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L18)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Multi-line body requires braces.
</details>
Done.

---

*[drake/systems/plants/xmlUtil.cpp, line 21 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIABjy_4gWCayUx2VDo:-KIDjf7c0G7NbLvmJNue:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L21)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Capitalize.
</details>
Done.

---

*[drake/systems/plants/xmlUtil.cpp, line 24 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAC66RlRVFnd_7P2iY:-KJ75fozWZ0FR5OJSYi-:-873310811) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L24)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> If strval has two items, this appears to return `true`, but should be `false`.  Missing unit test for this scenario.
</details>
Done. Added unit test called `test_parse_three_vector.string_input_two_vector_failure_mode`.

---

*[drake/systems/plants/xmlUtil.cpp, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAEjyIxmUF6oxfUYNM:-KJ76k7giAXhyRUUlUtQ:1718660787) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Should a strval with more than three items be a parse error?
</details>
Done. See new test `test_parse_three_vector.string_input_more_than_three_vector_failure_mode`.

---

*[drake/systems/plants/xmlUtil.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAEzNhjIxR2KyFrGGz:-KJ7Ld7-OgJPfeoeKvnz:2023918916) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test case for this branch.
</details>
Done. See new test `test_parse_three_vector.string_input_null_failure_mode`.

---

*[drake/systems/plants/xmlUtil.cpp, line 39 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAF2xRaUbiH3QFrDvU:-KJ7MTt8WqKMLhJnwuEw:472396364) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L39)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test case for this branch.
</details>
Done. See new test `test_parse_three_vector.node_input_nullptr`.

---

*[drake/systems/plants/xmlUtil.cpp, line 42 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAF9_tlK4Gg5E9NE4C:-KJ7NavQMrAInm28J7n-:612810354) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Isn't it weird that other functions treat a null node or strval as a parse error, but this one segfaults?
</details>
Done. I modified it to check whether `node` or the `element_name` are null and to return false if either are null. See new unit test `test_parse_three_vector.node_child_input_nullptr_failure_mode`.

---

*[drake/systems/plants/xmlUtil.cpp, line 50 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAFF7T_MVvTFctLjl1:-KJ7OYsqQ50_1PjIxFfa:-1089528767) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.cpp#L50)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Isn't it weird that other functions treat a null or strval node as a parse error, but this one segfaults?
</details>
Done. I modified it to check whether `node` or the `attribute_name` are null and to return false if either are null. See new unit test `test_parse_three_vector.node_attribute_input_nullptr`.

---

*[drake/systems/plants/xmlUtil.h, line 5 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA7UggE7dYftVPu8cF:-KJ7P-A-GFruSjJQu8FN:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> These two are in the wrong place now.
</details>
Done.

---

*[drake/systems/plants/xmlUtil.h, line 10 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA7azIAPvzZtPl3IXW:-KJ7P-wTT8omgHH8zjaH:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This one is in the wrong place now.
</details>
Done.

---

*[drake/systems/plants/xmlUtil.h, line 44 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA8PLaI3OrEz5BnXWv:-KJ7RElsaKkqkuSY0B5l:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> API doc fails to explain or cite the syntax(es) of the strval that will be successfully parsed.
</details>
Done.

---

*[drake/systems/plants/xmlUtil.h, line 52 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADOJA_nQB-1gwBe8a:-KJ7QWiAydjj62Yu87fm:861163732) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test for this API claim.  (Perhaps just remove the API claim?)
</details>
Done. See unit test `test_parse_three_vector.string_input_null_failure_mode`.

---

*[drake/systems/plants/xmlUtil.h, line 67 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADScpvUOFvkVOAEkx:-KJ7T5FYjaPFQoh2-9XC:2100743644) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L67)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test for this API claim.  (Perhaps just remove the API claim?)
</details>
Done. See unit test `test_parse_three_vector.node_input_nullptr`.

---

*[drake/systems/plants/xmlUtil.h, line 70 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIA9yuxlpX1q__o5drY:-KJ7TVOvbfCTiJaOUklm:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L70)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `const tinyxml2::XMLElement*` type for input
</details>
Done.

---

*[drake/systems/plants/xmlUtil.h, line 86 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADU3ZHcGm1ZJmpdkP:-KJ7U1AoX0GLVdq9Gris:639451418) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L86)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test for this API claim.  (Perhaps just remove the API claim?)
</details>
Done. See unit test `test_parse_three_vector.node_child_input_nullptr_failure_mode`.

---

*[drake/systems/plants/xmlUtil.h, line 89 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAA8qMV03baXGC3Pty:-KJ7TXaAv3fB9o-5C-Xp:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `const ...XMLElement*` type for input
</details>
Done.

---

*[drake/systems/plants/xmlUtil.h, line 107 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIADUmJneoG1P9JjXIY:-KJ7Twm8q-r9K3Mk-rN-:1690225465) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L107)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test for this API claim.  (Perhaps just remove the API claim?)
</details>
Done. See unit test `test_parse_three_vector.node_attribute_input_nullptr`.

---

*[drake/systems/plants/xmlUtil.h, line 110 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIAA9ZGBX-E4nwID4D1:-KJ7TYkv-OIeenGUwz7X:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/xmlUtil.h#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `const ...XMLElement*` type for input
</details>
Done.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 30 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIACiKodr2iAWr1IkUH:-KJ6jKTTR600PIjmxP8o:-2064649685) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/test/testXmlUtil.cpp#L30)):*
I looked through Google Test's manual and couldn't find any style guide stating that the first argument, which is the `test_case_name`, must equal the file name or that it must be the same across all tests within a single file. The following quote was taken from [this webpage](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#simple-tests):

> The first argument is the name of the test case, and the second argument is the test's name within the test case. Both names must be valid C++ identifiers, and they should not contain underscore (_). A test's full name consists of its containing test case and its individual name. Tests from different test cases can have the same individual name. ... Google Test groups the test results by test cases, so logically-related tests should be in the same test case; in other words, the first argument to their TEST() should be the same.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 34 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIACuGD9Vpj1B9wloVE:-KJ6kSg_gXci6wzPDPUX:-437080431) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/test/testXmlUtil.cpp#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW If Eigen weren't so fabulously verbose for how to zero things, it would be a defect for lack of initializing a variable at its point of declaration.
</details>
I added code that initializes `parsed_vector3d` to be a vector of zeros. (dishwasher safe)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJ7VgXI8fRf51E0cPui:701515998)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@jwnimmer-tri for final platform review / merge (following up from earlier review done two weeks ago).

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL. Thanks!
</details>


---

Review status: 5 of 8 files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJ7_HeWLuEmB_KvtQva:2054763715)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Just one item remains.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@jwnimmer-tri for final platform review / merge (following up from earlier review done two weeks ago).
</details>


---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 36 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJ8V_iTawuWfi6WxJop:-KJ8V_iTawuWfi6WxJoq:1603970444) ([raw file](https://github.com/robotlocomotion/drake/blob/325fc54f6fed909151cf9c131bb97be5d12a42c8/drake/systems/plants/xmlUtil.cpp#L36)):*
> ```cpp
>   ss >> token;
> 
>   (*val)[0] = std::stod(token);
> ```

BTW The stod will not error if the suffix is not-a-double.  For example `stod("1.23foo")` returns 1.23 and the "foo" part is sliently ignored.

In an ideal world, that would be a parse error (return false).  At some point Drake should use a principled XML parsing scheme to catch junk-on-numbers errors like that, but for now I think it's okay to only handle the "user typed the wrong count of items into the attribute" errors per the recent fixes here, and gloss over the accepts-garbage-strings errors for now.  So I'm fine with the PR as-is.  I'm just disappointed in general that translation from XML into structured data has to be crafted by hand like this, since it's so hard to get correct.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 30 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KIACiKodr2iAWr1IkUH:-KJ8ZHa2sVaUxVUjifns:1072229350) ([raw file](https://github.com/robotlocomotion/drake/blob/f29ad7a7df615f3ba01fb265046ec1b974ef92e8/drake/systems/plants/test/testXmlUtil.cpp#L30)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I looked through Google Test's manual and couldn't find any style guide stating that the first argument, which is the `test_case_name`, must equal the file name or that it must be the same across all tests within a single file. The following quote was taken from [this webpage](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#simple-tests):
> 
> > The first argument is the name of the test case, and the second argument is the test's name within the test case. Both names must be valid C++ identifiers, and they should not contain underscore (_). A test's full name consists of its containing test case and its individual name. Tests from different test cases can have the same individual name. ... Google Test groups the test results by test cases, so logically-related tests should be in the same test case; in other words, the first argument to their TEST() should be the same.
> 
</details>
I guess its okay.  It seemed a bit confusing that `testXmlUtil` was the filename and used for just the one test above, and then all of these tests have a different name.  But it is relatively internally consistent, so I guess it's fine.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 119 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJ8YecnGBL1vplDuWKR:-KJ8YecojzVPs95CT1uJ:1350355487) ([raw file](https://github.com/robotlocomotion/drake/blob/325fc54f6fed909151cf9c131bb97be5d12a42c8/drake/systems/plants/test/testXmlUtil.cpp#L119)):*
> ```cpp
>     EXPECT_THROW(
>         ParseThreeVectorValue(non_double_value_string[ii], &parsed_vector3d),
>         std::invalid_argument);
> ```

So, yay!  I'm glad this has a test now.

But I'm dubious that invalid numbers should be exceptions?  Why are some parse errors `false` and other parse errors `throw`?

The `runtime_errors` tested above seem like programming errors, not XML-content errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJ8ZnyA9HieoLJwvdR7:181083371)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@jwnimmer-tri I believe I addressed your comments. PTAL. Thanks!

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Just one item remains.
</details>


---

Review status: 5 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 36 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJ8V_iTawuWfi6WxJop:-KJBWbEGpNw-FXH-MPnG:-1025721839) ([raw file](https://github.com/robotlocomotion/drake/blob/325fc54f6fed909151cf9c131bb97be5d12a42c8/drake/systems/plants/xmlUtil.cpp#L36)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW The stod will not error if the suffix is not-a-double.  For example `stod("1.23foo")` returns 1.23 and the "foo" part is sliently ignored.
> 
> In an ideal world, that would be a parse error (return false).  At some point Drake should use a principled XML parsing scheme to catch junk-on-numbers errors like that, but for now I think it's okay to only handle the "user typed the wrong count of items into the attribute" errors per the recent fixes here, and gloss over the accepts-garbage-strings errors for now.  So I'm fine with the PR as-is.  I'm just disappointed in general that translation from XML into structured data has to be crafted by hand like this, since it's so hard to get correct.
</details>
Thanks for catching this. I modified the logic to detect extraneous characters after a double value in a string.

Maybe we should consider using https://github.com/leethomason/tinyxml2?

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 119 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJ8YecnGBL1vplDuWKR:-KJBXOv6X3m5RzC33eX2:-1641960426) ([raw file](https://github.com/robotlocomotion/drake/blob/325fc54f6fed909151cf9c131bb97be5d12a42c8/drake/systems/plants/test/testXmlUtil.cpp#L119)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> So, yay!  I'm glad this has a test now.
> 
> But I'm dubious that invalid numbers should be exceptions?  Why are some parse errors `false` and other parse errors `throw`?
> 
> The `runtime_errors` tested above seem like programming errors, not XML-content errors.
</details>
The returning of a bool was an attempt to being consistent with the existing methods in `xmlUtil.h`. Upon further reflection, I now think it's unnecessary to return a boolean and have modified all newly added `ParseThreeVectorValue()` methods to return void but throw an exception if it encounters any problem. I've also modified all thrown exceptions to be `std::invalid_argument` exceptions since, you're right, they are caused by programming errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJBaydBuQ2H4mNFNoyv:-726237442)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 55 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBgH9G-jgi0A-bGTz7:-KJBgH9H6m4L-n9EP09v:280210762) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/xmlUtil.cpp#L55)):*
> ```cpp
> 
>   ss >> token;
>   (*val)[1] = std::stod(token);
> ```

This stod is missing the num_chars check.  (I assume you wanted to catch all of them.)  A subroutine that wraps stod seems merited.

---

*[drake/systems/plants/xmlUtil.cpp, line 64 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBgRutNsfObkcM2zHw:-KJBgRutNsfObkcM2zHx:280210762) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/xmlUtil.cpp#L64)):*
> ```cpp
> 
>   ss >> token;
>   (*val)[2] = std::stod(token);
> ```

This stod is missing the num_chars check.  (I assume you wanted to catch all of them.)  A subroutine that wraps stod seems merited.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 104 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBfeC1j-8sJTOtFqv6:-KJBfeC1j-8sJTOtFqv7:935768356) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/test/testXmlUtil.cpp#L104)):*
> ```cpp
> GTEST_TEST(test_parse_three_vector,
>            string_input_non_double_vector_failure_mode) {
>   const int num_value_strings = 7;
> ```

Oops, I just noticed this.  It's a bit brittle to require editing the array contents + length separately (they could get out of sync).  I think `vector<string>` for example  would work fine and remove the brittleness.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 245 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBfyOAr88bW1fKAA8I:-KJBfyOAr88bW1fKAA8J:-621809277) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/test/testXmlUtil.cpp#L245)):*
> ```cpp
>   parsed_vector3d << 0, 0, 0;
> 
>   EXPECT_THROW(ParseThreeVectorValue(nullptr, "bar", &parsed_vector3d), std::invalid_argument);
> ```

META This line is too long now (as are several others in this file).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I believe I've addressed all reviewer comments. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@jwnimmer-tri I believe I addressed your comments. PTAL. Thanks!
</details>


---

Review status: 5 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 55 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBgH9G-jgi0A-bGTz7:-KJCCoywXJITutCmceO5:-1467214451) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/xmlUtil.cpp#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This stod is missing the num_chars check.  (I assume you wanted to catch all of them.)  A subroutine that wraps stod seems merited.
</details>
Done. See newly added method `StringToDouble(...)`. Should this method be added to `drake/common/`?

---

*[drake/systems/plants/xmlUtil.cpp, line 64 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBgRutNsfObkcM2zHw:-KJCCcKZHjornhWBaUGb:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/xmlUtil.cpp#L64)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This stod is missing the num_chars check.  (I assume you wanted to catch all of them.)  A subroutine that wraps stod seems merited.
</details>
Done.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 104 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBfeC1j-8sJTOtFqv6:-KJCEhthsYR92NVuPHbh:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/test/testXmlUtil.cpp#L104)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Oops, I just noticed this.  It's a bit brittle to require editing the array contents + length separately (they could get out of sync).  I think `vector<string>` for example  would work fine and remove the brittleness.
</details>
Done.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 245 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBfyOAr88bW1fKAA8I:-KJCIlmUtl5LPBvcncpQ:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/test/testXmlUtil.cpp#L245)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META This line is too long now (as are several others in this file).
</details>
Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJCIp_UZ0PjG2KMcp1i:-1596545867)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 55 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJBgH9G-jgi0A-bGTz7:-KJCJs66E5np9m4dNBgA:-1568190086) ([raw file](https://github.com/robotlocomotion/drake/blob/1ac8def47adf59562d914cfb77181bb362016c8f/drake/systems/plants/xmlUtil.cpp#L55)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Done. See newly added method `StringToDouble(...)`. Should this method be added to `drake/common/`?
</details>
I think its fine leaving it in XmlUtil until it's more widely needed.

---

*[drake/systems/plants/xmlUtil.h, line 133 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJCJ2yAkTdyy_Gay8xr:-KJCJ2yAkTdyy_Gay8xs:-739674001) ([raw file](https://github.com/robotlocomotion/drake/blob/5d240243a36e0d196a1d23ec39f3fd98386f610a/drake/systems/plants/xmlUtil.h#L133)):*
> ```objc
>  */
> DRAKEXMLUTIL_EXPORT
> double StringToDouble(const std::string* str);
> ```

Prefer to pass temporary values by const ref, not by const pointer.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 117 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJCJeIMmyqE08BArBTf:-KJCJeINwpMNP31Coe87:-1533631990) ([raw file](https://github.com/robotlocomotion/drake/blob/5d240243a36e0d196a1d23ec39f3fd98386f610a/drake/systems/plants/test/testXmlUtil.cpp#L117)):*
> ```cpp
>   parsed_vector3d << 0, 0, 0;
> 
>   for (std::string& bad_value : invalid_strings) {
> ```

Missing const on `bad_value`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Looks like I broke some other unit tests by making `ParseThreeVectorValue()` throw an exception with every error. I will look into fixing this soon.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I believe I've addressed all reviewer comments. PTAL.
</details>


---

Review status: 5 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.h, line 133 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJCJ2yAkTdyy_Gay8xr:-KJCZe-4sDFWlEqfPAyC:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5d240243a36e0d196a1d23ec39f3fd98386f610a/drake/systems/plants/xmlUtil.h#L133)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Prefer to pass temporary values by const ref, not by const pointer.
</details>
Done.

---

*[drake/systems/plants/test/testXmlUtil.cpp, line 117 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJCJeIMmyqE08BArBTf:-KJCZidXWnOyfmWKvSsb:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5d240243a36e0d196a1d23ec39f3fd98386f610a/drake/systems/plants/test/testXmlUtil.cpp#L117)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing const on `bad_value`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJC_QxFiZdL5_zNRGyZ:-1203146962)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I believe I fixed the breakage in various unit tests due to modifying `ParseThreeVectorAttribute()` to always throw an exception if it encounters an error. Hopefully the CI tests will pass this time.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Looks like I broke some other unit tests by making `ParseThreeVectorValue()` throw an exception with every error. I will look into fixing this soon.
</details>


---

Review status: 5 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJDEmLrQwdyWmGfrewi:-975336174)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/plants/xmlUtil.cpp, line 112 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJDF-JHlydk0Mzceemq:-KJDF-JHlydk0Mzceemr:1245358873) ([raw file](https://github.com/robotlocomotion/drake/blob/47140782cc9d953fef5dfaf104f6567bbff623f6/drake/systems/plants/xmlUtil.cpp#L112)):*
> ```cpp
>     const char* scale = node->Attribute(attribute_name);
>     if (scale == nullptr) {
>           throw std::invalid_argument(
> ```

The indentation seems wrong here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/xmlUtil.cpp, line 112 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-KJDF-JHlydk0Mzceemq:-KJDFQnDgepoACTnVsse:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/47140782cc9d953fef5dfaf104f6567bbff623f6/drake/systems/plants/xmlUtil.cpp#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The indentation seems wrong here.
</details>
Fixed. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I believe I fixed the breakage in various unit tests due to modifying `ParseThreeVectorAttribute()` to always throw an exception if it encounters an error. Hopefully the CI tests will pass this time.
</details>


---

Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJDGE7E5_SP0QYdEpht:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/1916/ and is thus ready for final review / merge.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJDS54MAZ9zUDakGpZz:-118482276)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
The msvc-64 is still building.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/1916/ and is thus ready for final review / merge.
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJDSSwNkHDzQVbWTZES:-1592498904)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Oops! Spoke too soon.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The msvc-64 is still building.
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2260#-:-KJDSg5J4WzBoGC_yz52:247501900)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Using override-merge due to out-of-date-with-master.
